Enable object as parameter for css, prop, and attr assertions

panzarino opened the pull request:
<!-- 
Thanks for contributing!
Read our contribution guidelines here: 
https://github.com/cypress-io/cypress/blob/develop/.github/CONTRIBUTING.md 
-->

<!-- Example: "Closes #1234" -->

- Closes #7353

### User facing changelog

Now have the option to pass an object of values to the `attr`, `prop`, and `css` assertion methods.

### Additional details

This change was needed because previously passing an object would always make the assertion pass. An exception could be thrown, but actually being able to pass an object would be a better option.

### How has the user experience changed?

Take the following test for example:
```javascript
this.$div = $('<div style=\'display: none; position: absolute;\'>div</div>')

// Group 1 - should pass
expect(this.$div).to.have.css({ display: 'none' })
expect(this.$div).to.have.css({ position: 'absolute' })
expect(this.$div).to.have.css({ display: 'none', position: 'absolute' })

// Group 2 - should not pass
expect(this.$div).to.have.css({ display: 'inline' })
expect(this.$div).to.have.css({ position: 'fixed' })
expect(this.$div).to.have.css({ display: 'inline', position: 'fixed' })
expect(this.$div).to.have.css({ foo: 'bar' })

// Group 3 - should pass
expect(this.$div).not.to.have.css({ display: 'inline' })
expect(this.$div).not.to.have.css({ position: 'fixed' })
expect(this.$div).not.to.have.css({ display: 'inline', position: 'fixed' })
expect(this.$div).not.to.have.css({ foo: 'bar' })
```

The intended behavior is that groups 1 and 3 pass, while group 2 does not.

Before these changes, groups 1 and 2 would pass while group 3 would not. The screenshot also shows that the correct comparison was not being done:
![Screen Shot 2020-05-14 at 8 36 45 PM](https://user-images.githubusercontent.com/7033952/81999290-aa64ab00-9622-11ea-8c7d-d945eacfee46.png)

After these changes, the correct comparisons are done and group 1 passes while group 2 does not. The screenshot shows how the correct comparison is used:
![Screen Shot 2020-05-14 at 8 38 41 PM](https://user-images.githubusercontent.com/7033952/81999397-ee57b000-9622-11ea-9152-73beeca520af.png)

As intended, groups 1 and 3 both pass with the correct comparisons being done:
![Screen Shot 2020-05-14 at 8 40 12 PM](https://user-images.githubusercontent.com/7033952/81999475-25c65c80-9623-11ea-865a-8365f02ce8e2.png)

### PR Tasks

<!-- 
These tasks must be completed before a PR is merged.
Delete tasks if they are not applicable. 
-->

- [x] Have tests been added/updated?
- [ ] Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [x] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?

Docs changes: https://github.com/cypress-io/cypress-documentation/pull/2803

panzarino committed:
Enable object as parameter for css, prop, and attr assertions

panzarino committed:
Merge branch 'develop' into issue-7353

panzarino committed:
Fill out tests for attr and prop

panzarino committed:
Merge branch 'develop' into issue-7353

panzarino committed:
Update ts type definitions

panzarino committed:
Merge branch 'issue-7353' of github.com:cypress-io/cypress into issue-7353

panzarino committed:
Merge branch 'develop' into issue-7353

jennifer-shehane commented on packages/driver/test/cypress/integration/commands/assertions_spec.js:
```diff
@@ -2350,7 +2350,7 @@ describe('src/cy/commands/assertions', () => {
     context('attr', () => {
       beforeEach(function () {
-        this.$div = $('<div foo=\'bar\'>foo</div>')
+        this.$div = $('<div foo=\'bar\' fizz=\'buzz\'>foo</div>')
```
You can use double quotes here - these escaped single quote `\'` are just an artifact of converting from CoffeeScript. I'd prefer double quotes.

jennifer-shehane reviewed (changes requested):
### 1. Assertion display differences

The display of the assertions differs when you use Cypress `.should()` assertions versus the standard expect. When using `.should()` it always only displays the last assertion in the Command Log. You should be able to write tests around the number/value of logs displayed in this situation.

My code against `<div style="display: block; position: absolute;">div</div>`

```js
it('visit', () => {
  cy.visit('index.html')
  cy.get('div').then(($div) => {
    expect($div).to.have.css({ display: 'block' })
    expect($div).to.have.css({ position: 'absolute' })
    expect($div).to.have.css({ display: 'block', position: 'absolute' })
  })
})

it('visit', () => {
  cy.visit('index.html')
  cy.get('div').should('have.css', { display: 'block' })
  cy.get('div').should('have.css', { position: 'absolute' })
  cy.get('div').should('have.css', { display: 'block', position: 'absolute' })
})
```

<img width="656" alt="Screen Shot 2020-05-20 at 12 36 54 PM" src="https://user-images.githubusercontent.com/1271364/82410910-be7f3100-9a96-11ea-9f04-344700d8abb0.png">

### 2. Empty object behavior

The assertions always pass when passed an empty Object. I dunno, maybe this is ok? Seems a bit weird though. The assertions don't display in the UI at all, so it seems like a noop.

```js
it('visit', () => {
  cy.visit('index.html')
  cy.get('div').then(($div) => {
    expect($div).to.have.css({})
  })
})

it('visit', () => {
  cy.visit('index.html')
  cy.get('div').should('have.css', {})
})
```

<img width="254" alt="Screen Shot 2020-05-20 at 12 41 42 PM" src="https://user-images.githubusercontent.com/1271364/82411194-4f560c80-9a97-11ea-9b43-5b6cb0601e19.png">

panzarino committed:
Update test

panzarino committed:
Merge branch 'develop' into issue-7353

panzarino committed:
Update type definitions

Bkucera commented on cli/types/index.d.ts:
```diff
@@ -1,33 +1,33 @@
+/// <reference path="./cypress.d.ts" />
+/// <reference path="./cypress-global-vars.d.ts" />
+/// <reference path="./cypress-type-helpers.d.ts" />
+/// <reference path="./cypress-expect.d.ts" />
```
```suggestion
/// <reference path="./cypress-expect.d.ts" />

```

panzarino commented:
@chrisbreiding and I decided that its probably best to not make these changes as they are fairly inconsistent with other methods and the existing Chai api, plus there would have to be a unique workaround to fix the logging issues. For now, I'm just going to submit another PR that causes a test to fail when passed an object rather than blindly succeed.

brian-mann commented:
I am not convinced this should not be implemented... I don't believe there are logging issues with this.

panzarino committed:
Merge branch 'develop' into issue-7353

panzarino committed:
Add (failing) tests around correct number of logs

panzarino committed:
Merge branch 'develop' into issue-7353

panzarino committed:
Update cli/types/index.d.ts

Co-authored-by: Ben Kucera <14625260+Bkucera@users.noreply.github.com>

jennifer-shehane commented:
Perhaps this illustrates the issue more clearly. When using `.should()`, only the last assertion in the object is logged as opposed to logging all assertions within a normal expect.

@brian-mann Can you explain more clearly why you consider this not an issue?

<img width="1027" alt="Screen Shot 2020-06-12 at 11 27 49 AM" src="https://user-images.githubusercontent.com/1271364/84466624-cc902e00-ac9f-11ea-8c24-dc9b4c662153.png">

Also, my second point about asserting on an empty object hasn't been responded to - whether this is desired behavior or not.

brian-mann commented:
With the object literal pattern - its really a "subset" kind of assertion, which means that the key's of the object are what you're asserting against.

If given a blank object then I believe this would give you a specific error telling you to provide **some** kind of assertion otherwise there's nothing specifically we can assert about.

panzarino committed:
Merge branch 'develop' into issue-7353

panzarino committed:
Simplify tests

brian-mann committed:
fixes inserting a command as a child when the assertion is outside of a command and the previous command in a hook was a cy.then

brian-mann committed:
prevent logging assertions multiple times when using expect(...)

- capture the assertionId and pass it around as state
- if the same underlying assertion is generating multiple chai
assertions, then merge the new assertions on the existing log instead
of generating multiple command logs

jennifer-shehane commented:
Closing due to inactivity. We can reopen a new PR if we feel the work is something we want to prioritize in the future.
